### PR TITLE
docs(flutter_weather): update tutorial

### LINKED ISSUE #2357: docs(flutter_weather): update documentation

### PR DESCRIPTION
**Pull request by jonathankao97**
## Status

in development

## Breaking Changes

no

## Description

update flutter weather tutorial (closes #2357)

## Type of Change

- [ ] ✨ New feature (non-breaking change which adds functionality)
- [ ] 🛠️ Bug fix (non-breaking change which fixes an issue)
- [ ] ❌ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🧹 Code refactor
- [ ] ✅ Build configuration change
- [x] 📝 Documentation
- [ ] 🗑️ Chore

### CODE REVIEW

File: docs/flutterweathertutorial.md [hunk 1]
```diff
@@ -2,39 +2,63 @@
 
 ![advanced](https://img.shields.io/badge/level-advanced-red.svg)
 
-> In the following tutorial, we're going to build a Weather app in Flutter which demonstrates how to manage multiple blocs to implement dynamic theming, pull-to-refresh, and much more. Our weather app will pull real data from an API and demonstrate how to separate our application into three layers (data, business logic, and presentation).
+## Tutorial Goals
+
+> The goal of this tutorial is to demonstrate best practices for design and testing when using the Bloc Library for a complex app. Specifically, we will use Hydrated Cubits to make a weather app.
```
Comment (felangel) [inline docs/flutterweathertutorial.md:7]:
```suggestion
> The goal of this tutorial is to demonstrate best practices for design and testing when using the bloc library to build a functional weather app. We will create a layered architecture, use `HydratedCubit`, and write unit and widget tests along the way.
```

File: docs/flutterweathertutorial.md [hunk 2]
```diff
@@ -2,39 +2,63 @@
...
 
 ![demo](./assets/gifs/flutter_weather.gif)
 
-## Key Topics
+### Project Requirements:
+- User can search for US cities on the search page
+- App displays weather information returned by [MetaWeather API](https://www.metaweather.com/api/)
+- App theme changes depending on weather of the city (Dynamic Theming)
+- Settings page which allows for unit conversions
+- Persist state across sessions (Hydrated Cubits)
```
Comment (felangel) [inline docs/flutterweathertutorial.md:16]:
```suggestion
- Persist state across sessions (HydratedCubit)
```

File: docs/flutterweathertutorial.md [hunk 3]
```diff
@@ -2,39 +2,63 @@
...
 - Observe state changes with [BlocObserver](/coreconcepts?id=blocobserver).
 - [BlocProvider](/flutterbloccoreconcepts?id=blocprovider), Flutter widget which provides a bloc to its children.
 - [BlocBuilder](/flutterbloccoreconcepts?id=blocbuilder), Flutter widget that handles building the widget in response to new states.
-- Using Bloc instead of Cubit. [What's the difference?](/coreconcepts?id=cubit-vs-bloc)
 - Prevent unnecessary rebuilds with [Equatable](/faqs?id=when-to-use-equatable).
 - [RepositoryProvider](/flutterbloccoreconcepts?id=repositoryprovider), a Flutter widget which provides a repository to its children.
 - [BlocListener](/flutterbloccoreconcepts?id=bloclistener), a Flutter widget which invokes the listener code in response to state changes in the bloc.
 - [MultiBlocProvider](/flutterbloccoreconcepts?id=multiblocprovider), a Flutter widget that merges multiple BlocProvider widgets into one.
 - [BlocConsumer](/flutterbloccoreconcepts?id=blocconsumer), exposes a builder and listener in order to react to new states.
+- [Hydrated Cubits](https://github.com/felangel/bloc/tree/master/packages/hydrated_bloc) to manage and mutate state, in response to user interactions
```
Comment (felangel) [inline docs/flutterweathertutorial.md:27]:
```suggestion
- [HydratedCubit](https://github.com/felangel/bloc/tree/master/packages/hydrated_bloc) to manage and mutate state, in response to user interactions
```

File: docs/flutterweathertutorial.md [hunk 4]
```diff
@@ -2,39 +2,63 @@
...
+- [Hydrated Storage](https://github.com/felangel/bloc/tree/master/packages/hydrated_bloc) to persist state
```
Comment (felangel) [inline docs/flutterweathertutorial.md:28]:
```suggestion
- [HydratedBloc](https://github.com/felangel/bloc/tree/master/packages/hydrated_bloc) to manage and persist state
```

File: docs/flutterweathertutorial.md [hunk 5]
```diff
@@ -2,39 +2,63 @@
...
+  |-- weather/
+  |-- main.dart
+|-- test/
+```
 
-and then install all of our dependencies
+### Architecture Layers
+> Following the [bloc architecture](https://bloclibrary.dev/#/architecture) guidelines, each feature will be implemented in layers. 
+- Data Layer: retrieving weather data from API
+- Repository: wrapping the data layer and compiling the data into a useful state for our application
```
Comment (felangel) [inline docs/flutterweathertutorial.md:56]:
```suggestion
- Repository: abstracting the data layer and exposing domain models for our application to consume
```

File: docs/flutterweathertutorial.md [hunk 6]
```diff
@@ -43,476 +67,2514 @@ We'll be focusing on two endpoints:
...
-Next, we need to allow users to get to the `Settings` widget from our `Weather` widget.
+Open [https://www.metaweather.com/api/location/search/?query=london](https://www.metaweather.com/api/location/search/?query=london) in your browser to see the response for the city of London. We will use the `woeid` (where-on-earth-id) in the return dictionary to hit the location endpoint.
+
+The `woeid` for London is 44418. Navigate to [https://www.metaweather.com/api/location/44418](https://www.metaweather.com/api/location/44418) in your browser and you'll see the response for weather in London. This contains all the data we will need for our app.
+
+### Creating Our MetaWeather API Client Subpackage
+
+> Since our API client interacting MetaWeather should be independent of the rest of our app, we can create it as a subpackage. We can then import our package in our `pubspec.yaml`
+
+Create a new directory on the project level called `packages`. This directory will store all of our custom made subpackages.
```
Comment (felangel) [inline docs/flutterweathertutorial.md:78]:
```suggestion
Create a new directory on the project level called `packages`. This directory will store all of our internal packages.
```

File: docs/flutterweathertutorial.md [hunk 7]
```diff
@@ -43,476 +67,2514 @@ We'll be focusing on two endpoints:
...
+
+Within this directory, run the built in flutter create command to create a new subpackage called `meta_weather_api` for our api client
```
Comment (felangel) [inline docs/flutterweathertutorial.md:80]:
```suggestion
Within this directory, run the built-in `flutter create` command to create a new package called `meta_weather_api` for our api client
```

File: docs/flutterweathertutorial.md [hunk 8]
```diff
@@ -43,476 +67,2514 @@ We'll be focusing on two endpoints:
...
+```dart
+// packages/meta_weather_api/lib/src/models/weather.dart (in progress)
+
+enum WeatherState {
+    snow, sleet, hail, thunderstorm, heavyRain, lightRain, showers, heavyCloud, lightCloud, clear, unknown
+}
+
+enum WindDirectionCompass {
+    north, northEast, east, southEast, south, southWest, west, northWest, unknown
+}
```
Comment (felangel) [inline docs/flutterweathertutorial.md:191]: can we format these to 80 characters so that developers don't have to scroll horizontally to see all values?

File: docs/flutterweathertutorial.md [hunk 9]
```diff
@@ -43,476 +67,2514 @@ We'll be focusing on two endpoints:
...
+|-- lib/
+|-- test/
+|-- packages/
+  |-- meta_weather_api/
+  |-- weather_repository
+    |-- lib/
+    |-- test/
+```
+
+Next, we will be use the same packages as in the `meta_weather_api` package. In addition, we will also use the `meta_weather_api` package which we just created. Update your `pubspec.yaml` to look like the following and run `flutter packages get`
```
Comment (felangel) [inline docs/flutterweathertutorial.md:1146]:
```suggestion
Next, we will be use the same packages as in the `meta_weather_api` package. In addition, we will also use the `meta_weather_api` package which we just created. Update your `pubspec.yaml` to look like the following and run `flutter pub get`
```

File: docs/flutterweathertutorial.md [hunk 10]
```diff
@@ -2,39 +2,63 @@
...
+To begin, create a new flutter project
+```bash
+flutter create flutter_weather
+```
 
-[pubspec.yaml](_snippets/flutter_weather_tutorial/pubspec.yaml.md ':include')
+### Project Structure: Feature Driven
+> Each package (directory) within our app contains an isolated feature. This allows for scalability as the number of features increases and also allows for developers to work on isolated parts of the codebase. 
 
-?> **Note:** We are going to add some assets (icons for weather types) in our app, so we need to include the assets folder in the pubspec.yaml. Please go ahead and create an _assets_ folder in the root of the project.
```
Comment (scarletteliza) [inline docs/flutterweathertutorial.md:31]: Can we add this note back in?

File: docs/flutterweathertutorial.md [hunk 11]
```diff
@@ -43,476 +67,2514 @@ We'll be focusing on two endpoints:
...
-?> **Note:** In the switch's `onChanged` method we add a `TemperatureUnitsToggled` event to notify the `SettingsBloc` that the temperature units have changed.
-
-Next, we need to allow users to get to the `Settings` widget from our `Weather` widget.
+Open [https://www.metaweather.com/api/location/search/?query=london](https://www.metaweather.com/api/location/search/?query=london) in your browser to see the response for the city of London. We will use the `woeid` (where-on-earth-id) in the return dictionary to hit the location endpoint.
+
+The `woeid` for London is 44418. Navigate to [https://www.metaweather.com/api/location/44418](https://www.metaweather.com/api/location/44418) in your browser and you'll see the response for weather in London. This contains all the data we will need for our app.
+
+### Creating Our MetaWeather API Client Subpackage
+
+> Since our API client interacting MetaWeather should be independent of the rest of our app, we can create it as a subpackage. We can then import our package in our `pubspec.yaml`
```
Comment (scarletteliza) [inline docs/flutterweathertutorial.md:76]: Can we provide a reason for this? Is it readability or a best practice? Maybe something like: 

We want to keep our API client independent of the rest of the app [for x reason(s)]... We can create it as a subpackage and then import it in our `pubspec.yaml`

File: docs/flutterweathertutorial.md (hunk 7 above)
Comment (scarletteliza) [inline docs/flutterweathertutorial.md:80]:
```suggestion
Within this directory, run the built-in flutter create command to create a new subpackage called `meta_weather_api` for our api client
```

File: docs/flutterweathertutorial.md [hunk 12]
```diff
@@ -2,39 +2,63 @@
...
+  |-- theme/
+  |-- weather/
+  |-- main.dart
+|-- test/
+```
 
-and then install all of our dependencies
+### Architecture Layers
+> Following the [bloc architecture](https://bloclibrary.dev/#/architecture) guidelines, each feature will be implemented in layers. 
+- Data Layer: retrieving weather data from API
```
Comment (scarletteliza) [inline docs/flutterweathertutorial.md:55]:
```suggestion
- Data Layer: retrieve weather data from API
```

File: docs/flutterweathertutorial.md (hunk 5 above)
Comment (scarletteliza) [inline docs/flutterweathertutorial.md:56]: Let's update the verb tense to be consistent with the other bullet points here

```suggestion
- Data Layer: retrieve weather data from API
- Repository: wrap the data layer and compiling the data into a useful state for our application
```

File: docs/flutterweathertutorial.md [hunk 13]
```diff
@@ -43,476 +67,2514 @@ We'll be focusing on two endpoints:
...
-
-We can then get the where-on-earth-id (woeid) and use it to hit the location api.
-
-Navigate to [https://www.metaweather.com/api/location/44418](https://www.metaweather.com/api/location/44418) in your browser and you'll see the response for weather in London. It should look something like this:
-
-[london.json](_snippets/flutter_weather_tutorial/location.json.md ':include')
-
-Great, now that we know what our data is going to look like, let’s create the necessary data models.
-
-## Creating Our Weather Data Model
```
Comment (scarletteliza) [inline docs/flutterweathertutorial.md:58]: Can we keep the spirit of these subheadings within the new tutorial? Let's break up the text into more bite-sized steps.

File: docs/flutterweathertutorial.md [hunk 14]
```diff
@@ -43,476 +67,2514 @@ We'll be focusing on two endpoints:
...
+
+Again, we will need to create our auto generated file using `build_runner`
+
+```
+flutter packages pub run build_runner build
+```
+
+### Theme Cubits
+
+Next, let's put together the HydratedCubit we will use to keep track of the theme of our app. The goal is to have the theme changed based on the current weather information.
```
Comment (scarletteliza) [inline docs/flutterweathertutorial.md:1973]:
```suggestion
Next, let's put together the HydratedCubit we will use to keep track of the theme of our app. The goal is to have the theme change based on the current weather information.
```

File: docs/flutterweathertutorial.md [hunk 15]
```diff
@@ -43,476 +67,2514 @@ We'll be focusing on two endpoints:
...
+```
+flutter_weather
+|-- lib/
+  |-- weather/
+    |-- theme/
+      |-- cubit/
+        |-- theme_cubit.dart
+```
+
+We will expose an `updateTheme(Weather? weather)` method, which will update the theme cubit depending on the weather condition
```
Comment (scarletteliza) [inline docs/flutterweathertutorial.md:1984]:
```suggestion
We will expose an `updateTheme(Weather? weather)` method, which will update the Theme Cubit depending on the weather condition
```

File: docs/flutterweathertutorial.md [hunk 16]
```diff
@@ -43,476 +67,2514 @@ We'll be focusing on two endpoints:
...
+```
+
+### Testing: Cubit Tests
+
+There are a few things we should test
+
+- If our weather states return correct `WeatherStatusX` values. [Here](https://github.com/felangel/bloc/blob/master/examples/flutter_weather/test/weather/cubit/weather_state_test.dart) is a completed test file
+- If our HydratedCubits respond to events properly and is able to convert fromJson and toJson
+
+In addition to using `Mocktail` for mocking and the standard testing library, it is also reccomended to use [bloc_test](https://pub.dev/packages/bloc_test) library. `bloc_test` allows us to easily prepare our blocs for testing, handle state changes, and check results.
```
Comment (scarletteliza) [inline docs/flutterweathertutorial.md:2040]:
```suggestion
In addition to using `Mocktail` for mocking and the standard testing library, it is also reccomended to use the [bloc_test](https://pub.dev/packages/bloc_test) library. `bloc_test` allows us to easily prepare our blocs for testing, handle state changes, and check results.
```

File: docs/flutterweathertutorial.md [hunk 17]
```diff
@@ -43,476 +67,2514 @@ We'll be focusing on two endpoints:
...
+
+[Here](https://github.com/felangel/bloc/blob/master/examples/flutter_weather/test/weather/cubit/weather_cubit_test.dart) is an example weather cubit testing file and [here](https://github.com/felangel/bloc/blob/master/examples/flutter_weather/test/theme/cubit/theme_cubit_test.dart) is an example theme cubit testing file
+
+## Presentation Layer: Weather Page, Settings Page, Search Page
+
+### Weather Page
+
+With the core logic of our app put together, it's time to put together the UI! We will start with the weather page which uses `dependency injection` in order to provide cubits to widgets. This is accomplished using `BlocProvider` and `BlocConsumer`. These objects essentially inject our Cubits whenever widgets are (re)rendered.
```
Comment (scarletteliza) [inline docs/flutterweathertutorial.md:2048]:
```suggestion
With the core logic of our app put together, it's time to put together the UI! We will start with the weather page which uses `dependency injection` in order to provide cubits to widgets. This is accomplished using `BlocProvider` and `BlocConsumer`. These objects inject our Cubits whenever widgets are (re)rendered.
```

File: docs/flutterweathertutorial.md [hunk 18]
```diff
@@ -43,476 +67,2514 @@ We'll be focusing on two endpoints:
...
+          final city = await Navigator.of(context).push(SearchPage.route());
+          unawaited(context.read<WeatherCubit>().fetchWeather(city));
+        },
+      ),
+    );
+  }
+}
+```
+
+This page depends on `SettingsPage, SearchPage` widgets, which we will create next
```
Comment (scarletteliza) [inline docs/flutterweathertutorial.md:2129]:
```suggestion
This page depends on `SettingsPage, SearchPage` widgets, which we will create next.
```

File: docs/flutterweathertutorial.md [hunk 19]
```diff
@@ -43,476 +67,2514 @@ We'll be focusing on two endpoints:
...
+        ],
+      ),
+    );
+  }
+}
+```
+
+## Presentation Layer: Individual Weather Widgets
+
+We will next create the UI for our app. Each individual widget corresponds to a particular state our Cubits can be in.
```
Comment (scarletteliza) [inline docs/flutterweathertutorial.md:2240]:
```suggestion
Now we will create the UI for our app. Each individual widget corresponds to a particular state that our Cubits can be in.
```

File: docs/flutterweathertutorial.md [hunk 20]
```diff
@@ -43,476 +67,2514 @@ We'll be focusing on two endpoints:
...
+        ),
+      ],
+    );
+  }
+}
+```
+
+### WeatherPopulated
+
+Next, lets create the UI for when a city has actually been queried for.
```
Comment (scarletteliza) [inline docs/flutterweathertutorial.md:2332]:
```suggestion
Next, let's create the UI for when we query a city.
```

File: docs/flutterweathertutorial.md [hunk 21]
```diff
@@ -43,476 +67,2514 @@ We'll be focusing on two endpoints:
...
+
+export 'weather_empty.dart';
+export 'weather_error.dart';
+export 'weather_loading.dart';
+export 'weather_populated.dart';
+```
+
+## Presentation Layer: Views
+
+Let's glue the rest of our app together!
```
Comment (scarletteliza) [inline docs/flutterweathertutorial.md:2484]:
```suggestion
Let's put the rest of our app together!
```

File: docs/flutterweathertutorial.md [hunk 22]
```diff
@@ -43,476 +67,2514 @@ We'll be focusing on two endpoints:
...
+        );
+      },
+    );
+  }
+}
+```
+
+## Testing Our Widgets and Views
+
+In order to feel comfortable about our testing coverage, we should also test the widgets and view pages we have created. Again, we also reccomend using the `bloc_test` library for this.
```
Comment (scarletteliza) [inline docs/flutterweathertutorial.md:2570]:
```suggestion
In order to feel confident about our testing coverage, we should also test the widgets and view pages we have created. Again, we recommend using the `bloc_test` library for this. 
```

File: docs/flutterweathertutorial.md [hunk 23]
```diff
@@ -43,476 +67,2514 @@ We'll be focusing on two endpoints:
...
+
+## Testing Our Widgets and Views
+
+In order to feel comfortable about our testing coverage, we should also test the widgets and view pages we have created. Again, we also reccomend using the `bloc_test` library for this. 
+
+The `bloc_test` library provides `MockBlocs` and `MockCubits` which make it easy to test UI based on particular states and events.
+
+[Here](https://github.com/felangel/bloc/tree/master/examples/flutter_weather/test) is an example test suite which includes widget tests, cubit tests, and view tests.
+
+## Putting it all together. Visually testing functionality
```
Comment (scarletteliza) [inline docs/flutterweathertutorial.md:2576]:
```suggestion
## Putting it all together!
```

File: docs/flutterweathertutorial.md [hunk 24]
```diff
@@ -43,476 +67,2514 @@ We'll be focusing on two endpoints:
...
+  http: ^0.13.0
+  json_annotation: ^4.0.0
+
+dev_dependencies:
+  coverage: ^0.14.1
+  build_runner: ^1.10.0
+  json_serializable: ^4.0.0
+```
+
+Makes sure to run `flutter pub get` after saving the file.
```
Comment (scarletteliza) [inline docs/flutterweathertutorial.md:567]:
```suggestion
Make sure to run `flutter pub get` after saving the file. 
```

File: docs/flutterweathertutorial.md [hunk 25]
```diff
@@ -43,476 +67,2514 @@ We'll be focusing on two endpoints:
...
+  if (weatherJson.isEmpty) {
+    throw WeatherRequestFailure();
+  }
+
+  return Weather.fromJson(weatherJson.first as Map<String, dynamic>);
+}
+```
 
-[weather.dart](_snippets/flutter_weather_tutorial/settings_button.dart.md ':include')
+The completed file, put together, looks like this
```
Comment (scarletteliza) [inline docs/flutterweathertutorial.md:662]:
```suggestion
Put together, the completed file should look like this:
```

File: docs/flutterweathertutorial.md [hunk 26]
```diff
@@ -43,476 +67,2514 @@ We'll be focusing on two endpoints:
...
+
+library meta_weather_api;
+
+export 'src/meta_weather_api_client.dart';
+export 'src/models/models.dart';
+```
+
+### Unit Testing our Data Layer Models
+
+First, let's unit test our data models, `location` and `weather`. We should focus on insuring that they can represent our data, translate from and to JSON form, and handle edge cases.
```
Comment (scarletteliza) [inline docs/flutterweathertutorial.md:749]:
```suggestion
First, let's unit test our data models, `location` and `weather`. We should focus on ensuring that they can represent our data, translate from and to JSON, and handle edge cases.
```

File: docs/flutterweathertutorial.md [hunk 27]
```diff
@@ -43,476 +67,2514 @@ We'll be focusing on two endpoints:
...
+
+First, add the Dart testing library to our `pubspec.yaml` and run `flutter pub get`.
```
Comment (scarletteliza) [inline docs/flutterweathertutorial.md:751]:
```suggestion
Add the Dart testing library to our `pubspec.yaml` and run `flutter pub get`.
```

File: docs/flutterweathertutorial.md [hunk 28]
```diff
@@ -43,476 +67,2514 @@ We'll be focusing on two endpoints:
...
+              .having((w) => w.visibility, 'visibility', 11.037727173307882)
+              .having((w) => w.predictability, 'predictability', 73),
+        );
+      });
+    });
+  });
+}
+```
+
+Finally, let's install the `coverage` package to run our tests with coverage and ensure that our tests pass
```
Comment (scarletteliza) [inline docs/flutterweathertutorial.md:1094]:
```suggestion
Finally, let's install the `coverage` package to run our tests with coverage and ensure that our tests pass.
```

File: docs/flutterweathertutorial.md [hunk 29]
```diff
@@ -43,476 +67,2514 @@ We'll be focusing on two endpoints:
...
+  test: ^1.16.4
+```
+
+```bash
+flutter test --coverage
+```
+
+## Repository Layer
+
+The goal of our repository layer is to create a wrapper for our data layer, and facilitate communication with the Bloc layer. In doing this, the rest of our code base depends only on functions exposed by our repository layer, instead of specific data provider implementations. This makes it easier for data providers to be switched out in the long run, and even converted into Flutter plugins/packages.
```
Comment (scarletteliza) [inline docs/flutterweathertutorial.md:1123]:
```suggestion
The goal of our repository layer is to create a wrapper for our data layer and facilitate communication with the Bloc layer. In doing this, the rest of our code base depends only on functions exposed by our repository layer, instead of specific data provider implementations. This makes it easier to switch out data providers later, or even convert them into individual Flutter plugins/packages.
```

File: docs/flutterweathertutorial.md [hunk 30]
```diff
@@ -43,476 +67,2514 @@ We'll be focusing on two endpoints:
...
+  coverage: ^0.14.1
+  build_runner: ^1.10.0
+  json_serializable: ^4.0.0
+  mocktail: ^0.1.0
+  test: ^1.16.4
+```
+
+### Creating Our Weather Repository Models
+
+Update your directory structure to look like this. We will be creating a new `weather.dart` models file to contain only data we want to store, and another `models.dart` barrel file.
```
Comment (scarletteliza) [inline docs/flutterweathertutorial.md:1174]:
```suggestion
Update your directory structure to look like the below. We will create a new `weather.dart` models file to contain only data we want to store, and another `models.dart` barrel file.
```

File: docs/flutterweathertutorial.md [hunk 31]
```diff
@@ -43,476 +67,2514 @@ We'll be focusing on two endpoints:
...
+}
+```
+
+```dart
+// packages/weather_repository/lib/src/models/models.dart (done)
+
+export 'weather.dart';
+```
+
+As we did in the previous section, run the following command to use `build_runner` to create our auto-generated models and factory methods.
```
Comment (scarletteliza) [inline docs/flutterweathertutorial.md:1236]:
```suggestion
As we did in the previous step, run the following command to use `build_runner` to create our auto-generated models and factory methods.
```

File: docs/flutterweathertutorial.md [hunk 32]
```diff
@@ -43,476 +67,2514 @@ We'll be focusing on two endpoints:
...
+      case WeatherState.lightCloud:
+        return WeatherCondition.cloudy;
+      default:
+        return WeatherCondition.unknown;
+    }
+  }
+}
+```
+
+Lastly, update the barrel file we created previously and we are done.
```
Comment (scarletteliza) [inline docs/flutterweathertutorial.md:1403]:
```suggestion
Lastly, update the barrel file we created previously, and we are done with this step!
```

File: docs/flutterweathertutorial.md [hunk 33]
```diff
@@ -43,476 +67,2514 @@ We'll be focusing on two endpoints:
...
+
+library weather_repository;
+
+export 'src/models/models.dart';
+export 'src/weather_repository.dart';
+```
+
+### Testing Our Weather Repository
+
+To test our Weather Repository, we will use the [Mocktail](https://github.com/felangel/mocktail) library. Using this library, we will create mocks and insure that our weather repository's `getWeather` method works as it should. This includes testing when the API client fails, and making sure the `Weather` object returned is correct when the API call is successful.
```
Comment (scarletteliza) [inline docs/flutterweathertutorial.md:1416]:
```suggestion
To test our Weather Repository, we will use the [Mocktail](https://github.com/felangel/mocktail) library. Using this library, we will create mocks and insure that our weather repository's `getWeather` method works as it should. This includes testing when the API client fails, and making sure the returned `Weather` object is correct when the API call is successful.
```

File: docs/flutterweathertutorial.md [hunk 34]
```diff
@@ -43,476 +67,2514 @@ We'll be focusing on two endpoints:
...
+  });
+}
+```
+
+
+
+
+## Business Logic Layer: Weather Model
+
+In this section, we're going to implement our third weather model - this time for the weather feature of our main app. Note that we haven't been implementing the exact same model each time. In the API client package, our weather model contained all the info returned by the API. In the repository client package, our weather model contained only the data we wish to display in our app, and in this weather model, we will bake temperature settings into the weather model.
```
Comment (scarletteliza) [inline docs/flutterweathertutorial.md:1586]:
```suggestion
In this section, we're going to implement our third weather model - this time for the weather feature of our main app. Note that we haven't been implementing the exact same model each time. In the API client package, our weather model contained all the info returned by the API. In the repository client package, our weather model contained only the data we wish to display in our app. In this weather model, we will bake temperature settings into the weather model.
```

File: docs/flutterweathertutorial.md [hunk 35]
```diff
@@ -43,476 +67,2514 @@ We'll be focusing on two endpoints:
...
+
+Before we get started, lets update our project level dependencies to import the packages we've created and also the serialization/deserialization dependencies.
```
Comment (scarletteliza) [inline docs/flutterweathertutorial.md:1588]:
```suggestion
Before we get started, let's update our project-level dependencies to import the packages we've created and also the serialization/deserialization dependencies.
```

File: docs/flutterweathertutorial.md [hunk 36]
```diff
@@ -43,476 +67,2514 @@ We'll be focusing on two endpoints:
...
+  build_runner: ^1.10.0
+  json_serializable: ^4.0.0
+
+flutter:
+  uses-material-design: true
+  assets:
+    - assets/
+```
+
+We will be working in the `weather` features app, in our main project library.
```
Comment (scarletteliza) [inline docs/flutterweathertutorial.md:1620]:
```suggestion
We will be working inside the `weather` features app, in our main project library.
```

File: docs/flutterweathertutorial.md [hunk 37]
```diff
@@ -43,476 +67,2514 @@ We'll be focusing on two endpoints:
...
+```
+flutter_weather
+|-- lib/
+  |-- weather/
+    |-- models/
+      |-- models.dart
+      |-- weather.dart
+```
+
+The goal of our weather model is to keep track of weather data displayed by our app and also to keep track of settings. To keep track of settings (celsius, farenheit), create the temperature model in `weather`
```
Comment (scarletteliza) [inline docs/flutterweathertutorial.md:1631]:
```suggestion
To keep track of settings (celsius, farenheit), create the temperature model in `weather`
```

Comment (scarletteliza) [reply docs/flutterweathertutorial.md:1631]: Moved the first sentence up to the top of this section

File: docs/flutterweathertutorial.md [hunk 38]
```diff
@@ -43,476 +67,2514 @@ We'll be focusing on two endpoints:
...
+      condition: condition ?? this.condition,
+      lastUpdated: lastUpdated ?? this.lastUpdated,
+      location: location ?? this.location,
+      temperature: temperature ?? this.temperature,
+    );
+  }
+}
+```
+
+Lastly, to wrap this section up, run `build_runner` to create the auto generated file and methods, and also export our models in the barrel file
```
Comment (scarletteliza) [inline docs/flutterweathertutorial.md:1755]:
```suggestion
To wrap this section up, run `build_runner` to create the auto generated file and methods, then export our models in the barrel file
```

File: docs/flutterweathertutorial.md [hunk 39]
```diff
@@ -43,476 +67,2514 @@ We'll be focusing on two endpoints:
...
+```
+
+```dart
+// lib/weather/models/models.dart (done)
+export 'weather.dart';
+```
+
+## Business Logic Layer: Weather and Theme Cubits
+
+In this tutorial, we will use HydratedCubits for our Bloc layer. HydratedCubits are essentially Cubits that can persist state across sessions when used in conjunction with HydratedStorage. In short, our HydratedCubits will keep track of state and expose methods which can be used to mutate state.
```
Comment (scarletteliza) [inline docs/flutterweathertutorial.md:1768]:
```suggestion
Now, we will use HydratedCubits for our Bloc layer to allow for theming. HydratedCubits are Cubits that can persist state across sessions when used in conjunction with HydratedStorage. In short, our HydratedCubits will keep track of state and expose methods which can be used to mutate state.
```

File: docs/flutterweathertutorial.md [hunk 40]
```diff
@@ -43,476 +67,2514 @@ We'll be focusing on two endpoints:
...
+
+extension WeatherStatusX on WeatherStatus {
+  bool get isInitial => this == WeatherStatus.initial;
+  bool get isLoading => this == WeatherStatus.loading;
+  bool get isSuccess => this == WeatherStatus.success;
+  bool get isFailure => this == WeatherStatus.failure;
+}
+```
+
+We will also need to keep track of what units (farenheit or celsius) our app is in.
```
Comment (scarletteliza) [inline docs/flutterweathertutorial.md:1798]:
```suggestion
We will also need to keep track of what units (Fahrenheit or Celsius) our app is using. 
```

File: docs/flutterweathertutorial.md [hunk 41]
```diff
@@ -43,476 +67,2514 @@ We'll be focusing on two endpoints:
...
+  Map<String, dynamic> toJson() => _$WeatherStateToJson(this);
+
+  @override
+  List<Object?> get props => [status, temperatureUnits, weather];
+}
+```
+
+### Weather Cubits
+
+Finally! It's time to put together our Weather Cubit. Our Weather Cubit will expose the following methods:
```
Comment (scarletteliza) [inline docs/flutterweathertutorial.md:1851]:
```suggestion
It's time to put together our Weather Cubit. Our Weather Cubit will expose the following methods:
```

**Review by felangel — APPROVED** [review]
LGTM 🚀 🎉
